### DISCUSSION TITLE
Merkle Proof over Mapping

### DISCUSSION THREAD

**Discussion by pierrealex3** [Cyfrin/foundry-full-course-cu#2867] [General]
Hello,

so I'm reaching out the end of the "Airdrop and Signatures" module of the Advanced Foundry course, and I haven't still figured out a clean answer about something that's been at the back of my mind... 

here's the deal:

Using a merkle proof mechanism to validate the *presence* of some data within a data structure has been the driver of this module.
It's started from the fact that "iterating a huge array of address containing the candidates for airdrop will not scale well".
And from what I see, a merkle proof mechanism has the following advantages for our scenario:
* it does scale well - the validation if quick!
* it does provide an advantage for storage costs: the merkle root stored as a state variable is small-sized: bytes32.

But it also has a dis-advantage: flexibility
If a user wanted to claim let's say only a fraction of the airdrop it's allowed - for whatever reason - the mechanism would not allow that, because the (account, amountToClaim) would not match the merkle root.

So my question...
In that case, wouldn't it be *as quick* and *more flexible* to simply opt for a mapping(address airdropCandidate => uint256 airdropBalance) ?
Would the disadvantage (and potential showstopper) be the size of the mapping as a storage variable?

please shed some light :)

**Comment by EngrPips** [comment]
Hello, @pierrealex3, Thinking about this `mapping` seems to be a great use-case for the situation you explained above, and obviously, it will come with the cost of updating storage variable every time a user claims a fraction of their token for the flexibility of partial claim that it allows I think the writing cost will be worth it for the protocol that wants to keep that feature. But one question that comes to my mind is how will you register all of the addresses that you want to airdrop token to the mappings

**Reply by pierrealex3** [reply]
"how will you register all of the addresses that you want to airdrop token to the mappings"
That is a very good point...
With my good'ol dev2 backend guy mindset I would say from the constructor, since this is data that needs to be provided upon deployment - or before the contract is useable, at least.

But then from what I've learned so far, providing a huge structure as an argument is a no-go, so it's not even an option in practice.
Then, the only other option I see is by providing an onlyOwner function that would allow updating the mapping one batch of entries at a time...  but then again, probably very gas costly for the sender!

So I could have answered my own question with this reflexion.  Merkle Proof over Mapping for a BIG difference in gas cost.
Do you see it the same way?

**Reply by EngrPips** [reply]
Exactly. Sometimes, we really want some feature or flexibility in how things work, but then the cost is unbearable, and we just have to let go. No one would be willing to bear that much cost just because they want users to claim part of their airdrop. I mean, they should claim everything at once. It will cause them no harm

**Reply by iLouisK** [reply]
Great question, I had similar thoughts in mind during the module and then completely forgot to explore the question and moved on - thanks for asking/answering!

**Reply by EngrPips** [reply]
Yeah, its a great question to ask.
